### ISSUE TITLE
Page for info that search engines can't easily find?

### ISSUE THREAD

**Issue by mars0i** [OCamlverse/ocamlverse.github.io#39]
I'm wondering whether it would be worth having a page--maybe under Learning, or under a top level link--just for links to information that's difficult to search for, usually because it involves non-alphabetic symbols used in unusual ways.  I'm not sure.  Just a thought.

Two examples: 

In `type +'a = ...`, `+` has a special meaning. I've found information on it it--searching on "type +" did work--but if one doesn't think of searching for "type +", there's no way it can be found.

I just came across `open!` in a code example--rather than `open`.  Still trying to find this one.

Maybe the internal search for ocamlverse would do better, but only if these things exist in the site.

**Comment by pmetzger** [comment]
Might be useful. Maybe we should have a FAQ?

**Comment by bluddy** [comment]
FAQ or 'OCaml subtleties' page sounds good to me. Should be linked from the main page IMO.

**Comment by loxs** [comment]
Two things that I often struggle to find via search engines:
* Docs for formatting options like `%s` `%b` etc.
* Docs for built-in infix operators like `@@` etc.

**Comment by bluddy** [comment]
'OCaml subtleties' makes more sense to me right now. FAQ doesn't necessarily suggest what we're looking to document here. I'd also like this page to talk about the main gotchas of OCaml: 1. polymorphic compare and 2. the semicolon problem in if statements.

@mars0i do you want to start adding this page? It can be linked from both the main page and learning.

**Comment by mars0i** [comment]
@bluddy Yes, I'll start it.

**Comment by mars0i** [comment]
I'm not sure I know the polymorphic compare problem, but I'll try to figure it out.  You mean the compare function e.g. in sorting?

**Comment by bluddy** [comment]
Yeah. Just mention it and I can fill it in.

**Comment by mars0i** [comment]
OK, will do.

**Comment by mars0i** [comment]
I'd be willing to start a separate FAQ page.  It would include a link to the subtleties page.  What do people think?  I'd suggest putting it on the main page.

My first contribution would be "What is a 'toplevel'?"  (This confused me when I first started with OCaml.  I knew what a REPL was, and wanted one.  I had no idea what a "toplevel" was.  The term is not intuitive.  Top level of what?  `utop` sounded like a REPL, but nothing I read called it that, so I wondered what the difference was.)

**Comment by mars0i** [comment]
@bluddy, is this the semicolon problem you had in mind:
```ocaml
# let f b =
  if b
  then let message = "It was true\n" in
    print_string message;
  print_string "All done!";;
val f : bool -> unit = <fun>
# f false;;
- : unit = ()
```
Uh oh. What happened to the "All done" message?

Here's the fix:
```ocaml
# let g b =
  if b
  then (let message = "It was true\n" in
    print_string message);
  print_string "All done!";;
val g : bool -> unit = <fun>
# g false;;
All done!- : unit = ()
```

**Comment by bluddy** [comment]
Yeah basically. But it's even worse than this, since a later edit can cause a `then` to become stateful.

**Comment by pmetzger** [comment]
BTW, important for some page that is easy to find: lists of common operators. Googling for operator names is hopeless.

**Comment by mars0i** [comment]
I agree about Googling.  The Ocaml Subtleties page has an operators entry with links to operators documentation pages in the manual.  That was one of the first entries added, because of the search engine problem.  

Do you think we should have an Operators link off the main page?

We could reproduce the documentation that's in the manual--I think a better job could be done--but to me that doesn't seem like a high priority, unless someone's just inspired to do it.  On the plus side for reproducing docs, operators won't change very quickly, which is a concern to have with duplicating language or library info that should be up to date elsewhere.

It might be worth having an operators link off the main page or as a separate link off the Learning page, even if all it is is a link to an anchor on the Subtleties page (or to wherever is that we put the links to the operator docs in the manual).

**Comment by mars0i** [comment]
I can add the semicolon text.

**Comment by mars0i** [comment]
Another thought.  At present, Google finds "ocaml subtleties" (Bing doesn't!), but "ocaml operators" doesn't find our page, presumably because there's too much separation between "ocaml" and "operators".  Should we preface all of our item entries with "ocaml" so that search engines will find them easily??  Or is there search engine metadata magic we can use?

**Comment by mars0i** [comment]
Interesting: "infix operator" in Google brings up HaskellWiki as the third link, and MLton a few entries down for that.  MLton but no OCaml!  And F# is the first entry on the second page.

**Comment by pmetzger** [comment]
> Do you think we should have an Operators link off the main page?

I think a FAQ with "common operators" and both links and a couple of things inline.

**Comment by mars0i** [comment]
This is kind of already on the Subtleties page.  More operators could be added.  But maybe there should be a separate operators FAQ?

**Comment by mars0i** [comment]
If we have a page called "OCaml operators", the search engines should find it.

**Comment by bluddy** [comment]
Some ideas to incorporate from [here](https://gist.github.com/jzelinskie/7609842). Good to see the point of view of an outsider, particularly things like 'array syntax is crazy' (which it kinda is).

**Comment by pmetzger** [comment]
I'd like to steal that gist wholesale if the author will let us. :)

**Comment by mars0i** [comment]
I wrote a semicolon problem explanation.  It's a simple issue but I ended up writing a long, more-friendly explanation, so I think it should have a separate page with a simple explanation in the Subtleties page linking to it.

**Comment by bluddy** [comment]
Sounds good. Try to give the file a specific name, like subtleties_semicolon_issue.md or something like that.

**Comment by mars0i** [comment]
I just added it, as [if_and_semicolon.md](content/if_and_semicolon.md), but I can rename it you prefer another name like the one you gave.  I don't have time to create the text in `subtleties.md` until later (but if someone else wants to add it before I do, that's fine).  Also feel free to edit the new file, of course.

**Comment by mars0i** [comment]
No problem.  If github restores it.  It 404'ed my edit save.

**Comment by mars0i** [comment]
Oh--@bluddy you renamed it while I was editing!  No problem.  It was just a tiny change.  But that's why I got a 404 when I saved.  :-)

**Comment by mars0i** [comment]
@bluddy, the if/semicolon problem seems a problem only when embedding a `let` in the final clause of the `if`.  Is it more general than that?  Whether there's an `else` or not, I can't seem to generate the problem without a `let`.

**Comment by mars0i** [comment]
After further experimentation, I think the if/semicolon issue--at least the one I described--is specifically an issue with `let` (and `if` and semicolon).  It's a consequence of `let`'s normal scoping rule.  It's just that when `let` is embedded in the last branch of an `if`, we think that the semicolon should end the `if`, but instead the semicolon is part of the `let` inside the `if`.  It would be strange if `let` behaved differently inside an `if`.  I added text to `subtleties_if_semicolon.md` that explains this.
